src/main/java/net/minecraftforge/unification/UnificationRegistry.java

```diff
@@ -0,0 +1,153 @@
+     */
+    UnificationRegistry(Function<T, ResourceLocation> namer, Function<T, T> validator, Function<T, T> outputCopier)
+    {
+        this.sortOrder = new UnifiedSortOrder<>(namer);
+        this.validator = validator;
+        this.outputCopier = outputCopier;
+    }
+
+    /**
+     * Add a variant to the registry.
+     * If no {@link Unified} exists yet, one is created.
+     * If a {@link Unified} already exist with these properties, the variant is added to its list of variants.
+     * For example usage, see {@link UnificationConstants#initVanillaEntries()}
+     */
+    public Unified<T> add(UnificationMaterial material, UnificationForm shape, T variant)
+    {
+        Preconditions.checkNotNull(material, "material must not be null");
+        Preconditions.checkNotNull(shape, "shape must not be null");
+
+        Map<UnificationForm, Unified<T>> map = UNIFICATION_TABLE.computeIfAbsent(material, k -> new IdentityHashMap<>());
+        Unified<T> unified = map.computeIfAbsent(shape, k -> new Unified<>(material, shape, outputCopier));
+        variant = validator.apply(variant);
+        unified.addVariant(variant, sortOrder);
+        return unified;
+    }
+
+    /**
+     * @return a {@link Unified} with the given properties.
+     * returns null if no {@link Unified} exists with those properties.
+     */
+    @Nullable
+    public Unified<T> get(UnificationMaterial material, UnificationForm shape)
+    {
+        Preconditions.checkNotNull(material, "material must not be null");
+        Preconditions.checkNotNull(shape, "shape must not be null");
+        Map<UnificationForm, Unified<T>> typeToUnified = UNIFICATION_TABLE.get(material);
+        if (typeToUnified == null) return null;
+        return typeToUnified.get(shape);
+    }
+
+    /**
+     * @return a map of all the {@link UnificationForm} registered for a given {@link UnificationMaterial}.
+     */
+    public Map<UnificationForm, Unified<T>> getShapesMapForMaterial(UnificationMaterial material)
+    {
+        Preconditions.checkNotNull(material, "material must not be null");
+        Map<UnificationForm, Unified<T>> map = UNIFICATION_TABLE.get(material);
+        if (map == null) return Collections.emptyMap();
+        return Collections.unmodifiableMap(map);
+    }
+
+    private static final class UnifiedSortOrder<T> implements Comparator<T>
+    {
+        private final Function<T, ResourceLocation> namer;
+
+        private UnifiedSortOrder(Function<T, ResourceLocation> namer)
+        {
+            this.namer = namer;
+        }
+
+        @Override
+        public int compare(T o1, T o2)
+        {
+            ResourceLocation registryName1 = namer.apply(o1);
+            ResourceLocation registryName2 = namer.apply(o2);
+
+            List<String> priorities = ForgeModContainer.unificationRegistryPriorityModIds;
+            int priorityIdx1 = priorities.indexOf(registryName1.getResourceDomain());
```

Comment:
Should we cache this?

Reply:
The list of priority mods is unlikely to be long at all. 
A pack maker might for example specify that they like TE the best, and then maybe BuildCraft next if they're being very specific. It should never be an expensive lookup.